### PR TITLE
hotfix/uci bpl build issues

### PR DESCRIPTION
**Pull request by vitalybu**
Fix build issues with the UCI BPL implementation and add the `beerocks` namespace to all BPL files.

### CODE REVIEW

File: framework/platform/bpl/linux/bpl.cpp [hunk 1]
```diff
@@ -17,17 +17,19 @@ SHARE_EASYLOGGINGPP(el::Helpers::storage())
...
 }
 
-void close()
+void bpl_close()
 {
     // Do nothing
 }
 
 } // namespace bpl
+} // namespace beerocks
```
Comment (tomereli) [inline framework/platform/bpl/linux/bpl.cpp]: Missing line break at end of file

File: framework/platform/bpl/linux/bpl_arp.cpp [hunk 2]
```diff
@@ -56,3 +55,4 @@ int arp_get_bridge_iface(const char bridge[BPL_ARP_IFACE_NAME_LEN],
 }
 
 } // namespace bpl
+} // namespace beerocks
```
Comment (tomereli) [inline framework/platform/bpl/linux/bpl_arp.cpp]: Missing line break at end of file

File: framework/platform/bpl/linux/bpl_dhcp.cpp [hunk 3]
```diff
@@ -17,3 +18,4 @@ int dhcp_mon_handle_event() { return 0; }
 int dhcp_mon_stop() { return 0; }
 
 } // namespace bpl
+} // namespace beerocks
```
Comment (tomereli) [inline framework/platform/bpl/linux/bpl_dhcp.cpp]: Missing line break at end of file

File: framework/platform/bpl/uci/arp/bpl_arp.cpp [hunk 4]
```diff
@@ -224,3 +223,4 @@ int arp_get_bridge_iface(const char bridge[BPL_ARP_IFACE_NAME_LEN],
 }
 
 } // namespace bpl
+} // namespace beerocks
```
Comment (tomereli) [inline framework/platform/bpl/uci/arp/bpl_arp.cpp]: Missing line break at end of file

File: framework/platform/bpl/uci/cfg/bpl_cfg_extra.cpp [hunk 5]
```diff
@@ -14,12 +14,10 @@
...
-
-using namespace beerocks::bpl;
-
+namespace beerocks {
 namespace bpl {
 
 int cfg_set_onboarding(int enable) { return 0; }
 
 } // namespace bpl
+} // namespace beerocks
```
Comment (tomereli) [inline framework/platform/bpl/uci/cfg/bpl_cfg_extra.cpp]: Missing line break at end of file

File: framework/platform/bpl/uci/cfg/bpl_cfg_helper.cpp [hunk 6]
```diff
@@ -94,3 +93,4 @@ int cfg_get_wep_key(int index, int keyIndex, char *key)
 }
 
 } // namespace bpl
+} // namespace beerocks
```
Comment (tomereli) [inline framework/platform/bpl/uci/cfg/bpl_cfg_helper.cpp]: Missing line break at end of file

File: framework/platform/bpl/uci/cfg/bpl_cfg_uci.cpp [hunk 7]
```diff
@@ -286,3 +289,4 @@ int cfg_uci_get_wireless_bool(enum paramType type, int index, const char param[]
 }
 
 } // namespace bpl
+} // namespace beerocks
```
Comment (tomereli) [inline framework/platform/bpl/uci/cfg/bpl_cfg_uci.cpp]: Missing line break at end of file

File: framework/platform/bpl/uci/dhcp/bpl_dhcp.cpp [hunk 8]
```diff
@@ -31,11 +34,12 @@ enum {
 };
 
 static const struct blobmsg_policy dhcp_event_policy[] = {
-    //TODO Remove: [DHCP_EVENT_ID] = { .name = "id", .type = BLOBMSG_TYPE_INT32 },
-    [DHCP_EVENT_OP]       = {.name = "op", .type = BLOBMSG_TYPE_STRING},
-    [DHCP_EVENT_MAC]      = {.name = "mac", .type = BLOBMSG_TYPE_STRING},
-    [DHCP_EVENT_IP]       = {.name = "ip", .type = BLOBMSG_TYPE_STRING},
-    [DHCP_EVENT_HOSTNAME] = {.name = "hostname", .type = BLOBMSG_TYPE_STRING}};
+    {.name = "id", .type = BLOBMSG_TYPE_INT32},       // [DHCP_EVENT_ID]
```
Comment (tomereli) [inline framework/platform/bpl/uci/dhcp/bpl_dhcp.cpp:37]: Shouldn't this be part of a separate commit?

File: framework/platform/bpl/uci/dhcp/bpl_dhcp.cpp [hunk 9]
```diff
@@ -96,24 +100,35 @@ static int dhcp_event_handler(struct ubus_context *ctx, struct ubus_object *obj,
     return 0;
 }
 
-static const struct ubus_method dhcp_ubus_methods[] = {
-    UBUS_METHOD("dhcp_event", dhcp_event_handler, dhcp_event_policy),
-};
+static const struct ubus_method dhcp_ubus_methods[] = {{.name     = "dhcp_event",
```
Comment (tomereli) [inline framework/platform/bpl/uci/dhcp/bpl_dhcp.cpp:103]: same question - shouldn't this be a separate commit?

File: tools/cmake/toolchain/openwrt.cmake [hunk 10]
```diff
@@ -48,5 +48,7 @@ set(CMAKE_C_IMPLICIT_INCLUDE_DIRECTORIES ${PLATFORM_TOOLCHAIN_DIR} CACHE STRING
 set(CMAKE_CXX_IMPLICIT_INCLUDE_DIRECTORIES ${PLATFORM_TOOLCHAIN_DIR} CACHE STRING "")
 
 ## Toolchain Programs
+set(CMAKE_AR ${PLATFORM_TOOLCHAIN_DIR}/bin/${PLATFORM_TOOLCHAIN_PREFIX}ar CACHE STRING "Archiver")
```
Comment (tomereli) [inline tools/cmake/toolchain/openwrt.cmake:51]: I would explain this a bit more in the commit message - this is NOT used right now, but once we switch to building statically we will need this for cross-compile.

**Review by tomereli — APPROVED** [review]
Mainly nitpicks so approving already since this is a critical fix for UGW.

File: framework/platform/bpl/uci/dhcp/bpl_dhcp.cpp (hunk 8 above)
Comment (vitalybu) [reply framework/platform/bpl/uci/dhcp/bpl_dhcp.cpp:37]: It's a tiny C/C++ build issue fix. Not critical.

File: framework/platform/bpl/uci/dhcp/bpl_dhcp.cpp (hunk 9 above)
Comment (vitalybu) [reply framework/platform/bpl/uci/dhcp/bpl_dhcp.cpp:103]: It's a tiny C/C++ build issue fix. Not critical.

File: framework/platform/bpl/linux/bpl.cpp (hunk 1 above)
Comment (vitalybu) [reply framework/platform/bpl/linux/bpl.cpp]: Fixed

File: framework/platform/bpl/linux/bpl_arp.cpp (hunk 2 above)
Comment (vitalybu) [reply framework/platform/bpl/linux/bpl_arp.cpp]: Fixed

File: framework/platform/bpl/linux/bpl_dhcp.cpp (hunk 3 above)
Comment (vitalybu) [reply framework/platform/bpl/linux/bpl_dhcp.cpp]: Fixed

File: framework/platform/bpl/uci/arp/bpl_arp.cpp (hunk 4 above)
Comment (vitalybu) [reply framework/platform/bpl/uci/arp/bpl_arp.cpp]: Fixed

File: framework/platform/bpl/uci/cfg/bpl_cfg_extra.cpp (hunk 5 above)
Comment (vitalybu) [reply framework/platform/bpl/uci/cfg/bpl_cfg_extra.cpp]: Fixed

File: framework/platform/bpl/uci/cfg/bpl_cfg_helper.cpp (hunk 6 above)
Comment (vitalybu) [reply framework/platform/bpl/uci/cfg/bpl_cfg_helper.cpp]: Fixed

File: framework/platform/bpl/uci/cfg/bpl_cfg_uci.cpp (hunk 7 above)
Comment (vitalybu) [reply framework/platform/bpl/uci/cfg/bpl_cfg_uci.cpp]: Fixed

File: agent/src/beerocks/slave/beerocks_slave_main.cpp [hunk 11]
```diff
@@ -293,7 +293,7 @@ static int run_beerocks_slave(beerocks::config_file::sConfigSlave &beerocks_slav
     //start platform_manager
     if (platform_mgr.init()) {
         // read the number of failures allowed before stopping agent from platform configuration
-        int stop_on_failure_attempts = bpl::cfg_get_stop_on_failure_attempts();
+        int stop_on_failure_attempts = beerocks::bpl::cfg_get_stop_on_failure_attempts();
```
Comment (arnout) [inline agent/src/beerocks/slave/beerocks_slave_main.cpp:296]: Is this needed? I thought because we are already in namespace beerocks, it wouldn't be needed...

File: agent/src/beerocks/slave/platform_manager/platform_manager_thread.cpp [hunk 12]
```diff
@@ -257,7 +257,7 @@ static bool fill_platform_settings(
 std::string extern_query_db(std::string parameter)
 {
     std::string ret;
-    if (bpl::init() < 0) {
+    if (beerocks::bpl::bpl_init() < 0) {
```
Comment (arnout) [inline agent/src/beerocks/slave/platform_manager/platform_manager_thread.cpp:260]: Why do you need to re-introduce the `bpl_` prefix? Why do this for bpl but not for bcl, tlvf, ieee1905_1, ...?

File: framework/platform/bpl/linux/bpl_cfg.cpp [hunk 13]
```diff
@@ -222,3 +221,4 @@ int cfg_notify_iface_status(const BPL_INTERFACE_STATUS_NOTIFICATION *status_noti
 int cfg_get_administrator_credentials(char pass[BPL_PASS_LEN]) { return RETURN_ERR; }
 
 } // namespace bpl
+} // namespace beerocks
```
Comment (arnout) [inline framework/platform/bpl/linux/bpl_cfg.cpp:224]: Doesn't clang-format fix the missing newline at the end of the file?

File: framework/platform/bpl/uci/cfg/bpl_cfg_uci.cpp [hunk 14]
```diff
@@ -8,7 +8,9 @@
 
 #include "bpl_cfg_uci.h"
 
+extern "C" {
 #include <uci.h>
+}
```
Comment (arnout) [inline framework/platform/bpl/uci/cfg/bpl_cfg_uci.cpp:13]: This is a completely unrelated change and should be a separate commit. It's something that is a really obvious required fix.

File: framework/platform/bpl/uci/dhcp/bpl_dhcp.cpp [hunk 15]
```diff
@@ -8,18 +8,21 @@
 
 #include <bpl/bpl_dhcp.h>
 
+extern "C" {
```
Comment (arnout) [inline framework/platform/bpl/uci/dhcp/bpl_dhcp.cpp:11]: Same here, should be in a separate commit that fixes the missing extern C includes.

File: framework/platform/bpl/uci/dhcp/bpl_dhcp.cpp (hunk 8 above)
Comment (arnout) [reply framework/platform/bpl/uci/dhcp/bpl_dhcp.cpp:37]: Begging to differ... It's two changes, which have nothing to do with the namespace change, and with no good explanation why they should be needed:

- it enables 'id' which wasn't enabled before and which isn't actually used, so there's no reason to add it;
- indexed initializers *are* supported by C++ AFAIK so there is no reason to not use them.

So in fact, it should be *two* separate commits, that explain why this change is needed.

File: framework/platform/bpl/uci/dhcp/bpl_dhcp.cpp [hunk 16]
```diff
@@ -68,16 +72,16 @@ static int dhcp_event_handler(struct ubus_context *ctx, struct ubus_object *obj,
                   blob_len(msg));
 
     if (tb[DHCP_EVENT_OP]) {
-        op = blobmsg_data(tb[DHCP_EVENT_OP]);
+        op = (const char *)blobmsg_data(tb[DHCP_EVENT_OP]);
```
Comment (arnout) [inline framework/platform/bpl/uci/dhcp/bpl_dhcp.cpp]: [Don't use C-style cast](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#es48-avoid-casts).

However, instead of `blobmsg_data()`, this should use `blobmsg_get_string()`.

File: framework/platform/bpl/uci/dhcp/bpl_dhcp.cpp (hunk 9 above)
Comment (arnout) [reply framework/platform/bpl/uci/dhcp/bpl_dhcp.cpp:103]: It's a huge change which I think is wrong, so yes it is critical. It absolutely requires a commit message that explain why UBUS_METHOD doesn't work.

File: framework/platform/bpl/uci/dhcp/bpl_dhcp.cpp [hunk 17]
```diff
@@ -96,24 +100,35 @@ static int dhcp_event_handler(struct ubus_context *ctx, struct ubus_object *obj,
...
-    .n_methods = ARRAY_SIZE(dhcp_ubus_methods),
+    .avl             = {},
+    .name            = "dhcp_event",
+    .id              = 0,
+    .path            = 0,
+    .type            = &dhcp_ubus_object_type,
+    .subscribe_cb    = 0,
+    .has_subscribers = 0,
+    .methods         = dhcp_ubus_methods,
+    .n_methods       = ARRAY_SIZE(dhcp_ubus_methods),
```
Comment (arnout) [inline framework/platform/bpl/uci/dhcp/bpl_dhcp.cpp:124]: Again, this shouldn't be needed.

File: tools/docker/build.sh [hunk 18]
```diff
@@ -17,9 +17,20 @@ main() {
...
+    "
+
+    # Add platform base directory mapping into the container
+    if [ ! -z "${PRPLMESH_PLATFORM_BASE_DIR}" ]; then
+        docker_args="${docker_args} \
+        -v ${PRPLMESH_PLATFORM_BASE_DIR}:${PRPLMESH_PLATFORM_BASE_DIR} \
+        "
+    fi
+
+    run docker container run ${docker_args} --rm prplmesh-builder $@
```
Comment (arnout) [inline tools/docker/build.sh:33]: Is this the only change needed to build in docker for a platform? That's nice. Only... why would you want to build in docker for a platform if the platform stuff is not in the docker image itself? That's the part I never understood about Dockerfile.openwrt...

**Review by arnout — CHANGES_REQUESTED** [review]
I believe this PR is changing too much, I think most of the changes are not needed to fix the bpl_uci build. I'll make a new PR with a simplified series that I think is sufficient.

File: agent/src/beerocks/slave/beerocks_slave_main.cpp (hunk 11 above)
Comment (vitalybu) [reply agent/src/beerocks/slave/beerocks_slave_main.cpp:296]: The main files are not in the beerocks namespace...

File: agent/src/beerocks/slave/platform_manager/platform_manager_thread.cpp (hunk 12 above)
Comment (vitalybu) [reply agent/src/beerocks/slave/platform_manager/platform_manager_thread.cpp:260]: Because only the BPL modules currently have a `close()` method that conflicts with the standard `close()` function form `unistd.h`

File: framework/platform/bpl/linux/bpl_cfg.cpp (hunk 13 above)
Comment (vitalybu) [reply framework/platform/bpl/linux/bpl_cfg.cpp:224]: Apparently not

File: framework/platform/bpl/uci/cfg/bpl_cfg_uci.cpp (hunk 14 above)
Comment (vitalybu) [reply framework/platform/bpl/uci/cfg/bpl_cfg_uci.cpp:13]: It's part of the necessary fixes after switching to C++ files...

File: framework/platform/bpl/uci/dhcp/bpl_dhcp.cpp (hunk 15 above)
Comment (vitalybu) [reply framework/platform/bpl/uci/dhcp/bpl_dhcp.cpp:11]: Again, required only after switching to CPP build

File: framework/platform/bpl/uci/dhcp/bpl_dhcp.cpp (hunk 17 above)
Comment (vitalybu) [reply framework/platform/bpl/uci/dhcp/bpl_dhcp.cpp:124]: It doesn't work because C++ doesn't support partial struct initializers (at least not the GCC version used in UGW). It doesn't change the functionality whatsoever.

File: tools/docker/build.sh (hunk 18 above)
Comment (vitalybu) [reply tools/docker/build.sh:33]: It's a temporary solution until we'll be able to package all target platforms into docker containers... Till then, I prefer to use a generic builder docker with a pre-built target tree.

File: framework/platform/bpl/uci/dhcp/bpl_dhcp.cpp (hunk 8 above)
Comment (vitalybu) [reply framework/platform/bpl/uci/dhcp/bpl_dhcp.cpp:37]: This change is required because C++ (at least the GCC version used in UGW) doesn't support reordered or partial initializer lists... It not that big of a change. Expanded the macro and set default values for non initialized fields.

File: framework/platform/bpl/uci/dhcp/bpl_dhcp.cpp (hunk 9 above)
Comment (vitalybu) [reply framework/platform/bpl/uci/dhcp/bpl_dhcp.cpp:103]: Why do you think it's a huge and wrong change?

File: framework/platform/bpl/uci/dhcp/bpl_dhcp.cpp (hunk 16 above)
Comment (vitalybu) [reply framework/platform/bpl/uci/dhcp/bpl_dhcp.cpp]: Right, missed that... Fixed.

**Comment by arnout** [comment]
Properly split in commits, with some comments added, and removing just a few things -> #600 

Here is the diff:

```
diff --git a/agent/src/beerocks/slave/platform_manager/platform_manager_thread.cpp b/agent/src/beerocks/slave/platform_manager/platform_manager_thread.cpp
index 5b44aabf..ae7e8342 100644
--- a/agent/src/beerocks/slave/platform_manager/platform_manager_thread.cpp
+++ b/agent/src/beerocks/slave/platform_manager/platform_manager_thread.cpp
@@ -257,7 +257,7 @@ static bool fill_platform_settings(
 std::string extern_query_db(std::string parameter)
 {
     std::string ret;
-    if (beerocks::bpl::bpl_init() < 0) {
+    if (bpl::bpl_init() < 0) {
         ret = "Failed to initialize BPL!";
     } else {
         if (parameter == "is_master") {
@@ -482,7 +482,7 @@ void main_thread::send_dhcp_notification(std::string op, std::string mac, std::s
 bool main_thread::init()
 {
     // Initialize the BPL (Beerocks Platform Library)
-    if (beerocks::bpl::bpl_init() < 0) {
+    if (bpl::bpl_init() < 0) {
         LOG(ERROR) << "Failed to initialize BPL!";
         return (false);
     }
@@ -679,7 +679,7 @@ void main_thread::on_thread_stop()
         m_pDHCPMonSocket = nullptr;
     }
 
-    beerocks::bpl::bpl_close();
+    bpl::bpl_close();
     LOG(DEBUG) << "Closed BPL.";
 
     bpl_iface_status_map.clear();
diff --git a/framework/platform/bpl/uci/dhcp/bpl_dhcp.cpp b/framework/platform/bpl/uci/dhcp/bpl_dhcp.cpp
index 6a7c000a..d17f0eb7 100644
--- a/framework/platform/bpl/uci/dhcp/bpl_dhcp.cpp
+++ b/framework/platform/bpl/uci/dhcp/bpl_dhcp.cpp
@@ -21,25 +21,19 @@ extern "C" {
 ////////////////////////// Local Module Definitions //////////////////////////
 //////////////////////////////////////////////////////////////////////////////
 
+namespace beerocks {
+namespace bpl {
+
 static struct ubus_context *s_pUbusCtx = NULL;
-static beerocks::bpl::dhcp_mon_cb s_pCallback = NULL;
-
-enum {
-    DHCP_EVENT_ID,
-    DHCP_EVENT_OP,
-    DHCP_EVENT_MAC,
-    DHCP_EVENT_IP,
-    DHCP_EVENT_HOSTNAME,
-    __DHCP_EVENT_MAX
-};
+static dhcp_mon_cb s_pCallback         = NULL;
+
+enum { DHCP_EVENT_OP, DHCP_EVENT_MAC, DHCP_EVENT_IP, DHCP_EVENT_HOSTNAME, __DHCP_EVENT_MAX };
 
 static const struct blobmsg_policy dhcp_event_policy[] = {
-    {.name = "id", .type = BLOBMSG_TYPE_INT32},       // [DHCP_EVENT_ID]
-    {.name = "op", .type = BLOBMSG_TYPE_STRING},      // [DHCP_EVENT_OP]
-    {.name = "mac", .type = BLOBMSG_TYPE_STRING},     // [DHCP_EVENT_MAC]
-    {.name = "ip", .type = BLOBMSG_TYPE_STRING},      // [DHCP_EVENT_IP]
-    {.name = "hostname", .type = BLOBMSG_TYPE_STRING} // [DHCP_EVENT_HOSTNAME]
-};
+    [DHCP_EVENT_OP]       = {.name = "op", .type = BLOBMSG_TYPE_STRING},
+    [DHCP_EVENT_MAC]      = {.name = "mac", .type = BLOBMSG_TYPE_STRING},
+    [DHCP_EVENT_IP]       = {.name = "ip", .type = BLOBMSG_TYPE_STRING},
+    [DHCP_EVENT_HOSTNAME] = {.name = "hostname", .type = BLOBMSG_TYPE_STRING}};
 
 struct dhcp_event_request {
     struct ubus_request_data req;
@@ -100,26 +94,17 @@ static int dhcp_event_handler(struct ubus_context *ctx, struct ubus_object *obj,
     return 0;
 }
 
-static const struct ubus_method dhcp_ubus_methods[] = {{.name     = "dhcp_event",
-                                                        .handler  = dhcp_event_handler,
-                                                        .mask     = 0,
-                                                        .tags     = 0,
-                                                        .policy   = dhcp_event_policy,
-                                                        .n_policy = ARRAY_SIZE(dhcp_event_policy)}};
+static const struct ubus_method dhcp_ubus_methods[] = {
+    UBUS_METHOD("dhcp_event", dhcp_event_handler, dhcp_event_policy),
+};
 
-static struct ubus_object_type dhcp_ubus_object_type = {.name      = "dhcp_event",
-                                                        .id        = 0,
-                                                        .methods   = dhcp_ubus_methods,
-                                                        .n_methods = ARRAY_SIZE(dhcp_ubus_methods)};
+// UBUS_OBJECT_TYPE puts the fields in the wrong order
+static struct ubus_object_type dhcp_ubus_object_type = {
+    .name = "dhcp_event", .methods = dhcp_ubus_methods, .n_methods = ARRAY_SIZE(dhcp_ubus_methods)};
 
 static struct ubus_object dhcp_ubus_object = {
-    .avl             = {},
     .name      = "dhcp_event",
-    .id              = 0,
-    .path            = 0,
     .type      = &dhcp_ubus_object_type,
-    .subscribe_cb    = 0,
-    .has_subscribers = 0,
     .methods   = dhcp_ubus_methods,
     .n_methods = ARRAY_SIZE(dhcp_ubus_methods),
 };
@@ -128,9 +113,6 @@ static struct ubus_object dhcp_ubus_object = {
 /////////////////////////////// Implementation ///////////////////////////////
 //////////////////////////////////////////////////////////////////////////////
 
-namespace beerocks {
-namespace bpl {
-
 int dhcp_mon_start(dhcp_mon_cb cb)
 {
     // Modules doesn't support multiple starts
```

**Comment by tomereli** [comment]
Obsolete
